src/syscheckd/create_db.c

```diff
@@ -85,49 +85,47 @@ void fim_scan() {
 #ifdef WIN32
         os_winreg_check();
 #endif
-
-    check_deleted_files();
-
     if (syscheck.file_limit_enabled) {
         w_mutex_lock(&syscheck.fim_entry_mutex);
         nodes_count = fim_db_get_count_entry_path(syscheck.database);
         w_mutex_unlock(&syscheck.fim_entry_mutex);
+    }
 
-        if (nodes_count < syscheck.file_limit) {
-            it = 0;
-
-            w_mutex_lock(&syscheck.fim_scan_mutex);
-
-            while ((syscheck.dir[it] != NULL) && (nodes_count < syscheck.file_limit)) {
-                struct fim_element *item;
-                os_calloc(1, sizeof(fim_element), item);
-                item->mode = FIM_SCHEDULED;
-                item->index = it;
-                fim_checker(syscheck.dir[it], item, NULL, 0);
-                it++;
-                os_free(item);
-
-                w_mutex_lock(&syscheck.fim_entry_mutex);
-                nodes_count = fim_db_get_count_entry_path(syscheck.database);
-                w_mutex_unlock(&syscheck.fim_entry_mutex);
-            }
+    check_deleted_files();
 
-            w_mutex_unlock(&syscheck.fim_scan_mutex);
+    w_mutex_lock(&syscheck.fim_entry_mutex);
+    fim_db_set_all_unscanned(syscheck.database);
+    w_mutex_unlock(&syscheck.fim_entry_mutex);
 
-#ifdef WIN32
-            if (nodes_count < syscheck.file_limit) {
-                os_winreg_check();
+    if (syscheck.file_limit_enabled && !syscheck.database->full && (nodes_count >= syscheck.file_limit)) {
+        it = 0;
 
-                w_mutex_lock(&syscheck.fim_entry_mutex);
-                fim_db_get_count_entry_path(syscheck.database);
-                w_mutex_unlock(&syscheck.fim_entry_mutex);
-            }
-#endif
+        w_mutex_lock(&syscheck.fim_scan_mutex);
+
+        while ((syscheck.dir[it] != NULL) && (!syscheck.database->full)) {
+            struct fim_element *item;
+            os_calloc(1, sizeof(fim_element), item);
+            item->mode = FIM_SCHEDULED;
+            item->index = it;
+            fim_checker(syscheck.dir[it], item, NULL, 0);
+            it++;
+            os_free(item);
 
             w_mutex_lock(&syscheck.fim_entry_mutex);
-            fim_db_set_all_unscanned(syscheck.database);
+            nodes_count = fim_db_get_count_entry_path(syscheck.database);
             w_mutex_unlock(&syscheck.fim_entry_mutex);
         }
+
+        w_mutex_unlock(&syscheck.fim_scan_mutex);
+
+#ifdef WIN32
+        if (nodes_count < syscheck.file_limit) {
```

Comment:
This condition should be replaced by checking the value of `syscheck.database->full` instead.